metals/src/main/scala/scala/meta/internal/metals/utils/FutureWithTimeout.scala

```diff
@@ -0,0 +1,41 @@
+package scala.meta.internal.metals.utils
+
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.Duration
+import scala.util.Failure
+import scala.util.Try
+
+import scala.meta.internal.metals.Cancelable
+import scala.meta.internal.metals.CancelableFuture
+import scala.meta.internal.metals.MetalsEnrichments._
+
+object FutureWithTimeout {
```

Comment:
I wonder if we could go another direction. Instead of relying on Await, we could save the request time and if another request comes in the queue and the timeout is calculated to be too much (using FlexTimeout) we could ask the user if they want to cancel it:
```
$name request is taking longer than expected, do you want to cancel and rerun it?
Cancel | Wait | Wait always
```

If they wait we could triple the flex time and ask again at some point. `Wait always` would imply that they don't want to show those pop ups, which could be safe if the feature is not 100 % at the start.